typing/typetexp.ml

```diff
@@ -1079,9 +1097,19 @@ and transl_type_aux env ~row_context ~aliased ~policy mode styp =
       let cty = transl_type new_env ~policy ~row_context mode t in
       ctyp (Ttyp_open (path, mod_ident, cty)) cty.ctyp_type
   | Ptyp_of_kind jkind ->
-    let tjkind = jkind_of_annotation (Type_of_kind loc) styp.ptyp_attributes jkind in
-    let ty = newty (Tof_kind tjkind) in
-    ctyp (Ttyp_of_kind jkind) ty
+      let tjkind =
+        jkind_of_annotation (Type_of_kind loc) styp.ptyp_attributes jkind
+      in
+      let ty = newty (Tof_kind tjkind) in
+      ctyp (Ttyp_of_kind jkind) ty
+  | Ptyp_quote t ->
+      let new_env = Env.add_quotation_lock env in
+      let cty = transl_type new_env ~policy ~row_context mode t in
+      ctyp (Ttyp_quote cty) (newty (Tquote cty.ctyp_type))
+  | Ptyp_splice t ->
+      let new_env = Env.add_splice_lock env in
```

Comment:
Bug here: this means writing `$(int -> int)` (outside of any quotation) would be allowed - but it shouldn't.

If we adopt my comment on `add_split_lock`, this bug should go away.

Can you add a test to prevent future regression?